File: Kepler-Server/src/main/java/org/alexdev/kepler/game/item/base/ItemDefinition.java
package org.alexdev.kepler.game.item.base;

import org.alexdev.kepler.game.texts.TextsManager;

import java.util.ArrayList;
import java.util.List;

public class ItemDefinition {
    public static final double DEFAULT_TOP_HEIGHT = 0.001;

    private int id;
    private String sprite;
    private String behaviourData;
    private double topHeight;
    private int length;
    private int width;
    private String colour;
    private List<ItemBehaviour> behaviourList;

    public ItemDefinition() {
        this.sprite = "";
        this.behaviourData = "";
        this.colour = "";
        this.topHeight = DEFAULT_TOP_HEIGHT;
        this.length = 1;
        this.width = 1;
        this.behaviourList = new ArrayList<>();
    }

    public ItemDefinition(int id, String sprite, String behaviourData, double topHeight, int length, int width, String colour) {
        this.id = id;
        this.sprite = sprite;
        this.behaviourData = behaviourData;
        this.topHeight = topHeight;
        this.length = length;
        this.width = width;
        this.colour = colour;
        this.behaviourList = parseBehaviour(this.behaviourData);

        // If the item is a gate (checked below) then the top height is set to 0 so the item can be walked in
        if (!this.behaviourList.contains(ItemBehaviour.CAN_SIT_ON_TOP)
                && !this.behaviourList.contains(ItemBehaviour.CAN_LAY_ON_TOP)
                && !this.behaviourList.contains(ItemBehaviour.CAN_STACK_ON_TOP)) {
            this.topHeight = 0;
        }

        // If the top height 0, then make it 0.001 to make it taller than the default room tile, that the
        // furni collision map can be generated.
        if (this.topHeight == 0) {
            this.topHeight = DEFAULT_TOP_HEIGHT;
        }
    }

    /**
     * Parse the behaviour list seperated by comma.
     *
     * @param behaviourData the behaviourData to parse
     * @return the behaviour list
     */
    private static List<ItemBehaviour> parseBehaviour(String behaviourData) {
        List<ItemBehaviour> behaviourList = new ArrayList<>();

        if (behaviourData.length() > 0) {
            for (String behaviourEnum : behaviourData.split(",")) {
                behaviourList.add(ItemBehaviour.valueOf(behaviourEnum.toUpperCase()));
            }
        }

        return behaviourList;
    }

    /**
     * Get if the item has a type of behaviour.
     *
     * @param behaviour the behaviour to check
     * @return true, if successful
     */
    public boolean hasBehaviour(ItemBehaviour behaviour) {
        return this.behaviourList.contains(behaviour);
    }

    /**
     * Add a behaviour to the list.
     *
     * @param behaviour the behaviour to add
     */
    public void addBehaviour(ItemBehaviour behaviour) {
        if (this.behaviourList.contains(behaviour)) {
            return;
        }

        this.behaviourList.add(behaviour);
    }

    /**
     * Remove a behaviour from the list.
     *
     * @param behaviour the behaviour to remove
     */
    public void removeBehaviour(ItemBehaviour behaviour) {
        this.behaviourList.remove(behaviour);
    }

    /**
     * Get the item name by creating an external text key and reading external text entries.
     *
     * @param specialSpriteId the special sprite id
     * @return the name
     */
    public String getName(int specialSpriteId) {
        if (this.hasBehaviour(ItemBehaviour.DECORATION)) {
            return this.sprite;
        }

        String etxernalTextKey = this.getExternalTextKey(specialSpriteId);
        String name = etxernalTextKey + "_name";

        String value = TextsManager.getInstance().getValue(etxernalTextKey);

        if (value == null) {
            return "null";
        }

        return name;
    }

    /**
     * Get the item description by creating an external text key and reading external text entries.
     *
     * @param specialSpriteId the special sprite id
     * @return the description
     */
    public String getDescription(int specialSpriteId) {
        if (this.hasBehaviour(ItemBehaviour.CAN_STACK_ON_TOP)) {
            return this.sprite;
        }

        String etxernalTextKey = this.getExternalTextKey(specialSpriteId);
        String name = etxernalTextKey + "_desc";

        String value = TextsManager.getInstance().getValue(etxernalTextKey);

        if (value == null) {
            return "null";
        }

        return name;
    }

    /**
     * Create the catalogue icon through using the special sprite id.
     *
     * @param specialSpriteId the special sprite id
     * @return the catalogue icon
     */
    public String getIcon(int specialSpriteId) {
        String icon = "";

        icon += this.sprite;

        if (specialSpriteId > 0) {
            icon += " " + specialSpriteId;
        }

        return icon;
    }

    /**
     * Get external text key by definition.
     *
     * @param specialSpriteId the special sprite id
     * @return the external text key
     */
    private String getExternalTextKey(int specialSpriteId) {
        String key = "";

        if (specialSpriteId == 0) {
            if (this.hasBehaviour(ItemBehaviour.WALL_ITEM)) {
                key = "wallitem";
            } else {
                key = "furni";
            }

            key += "_";
        }

        key += this.sprite;

        if (specialSpriteId > 0) {
            key += ("_" + specialSpriteId);
        }

        return key;
    }

    public int getId() {
        return id;
    }

    public String getSprite() {
        return sprite;
    }

    public void setSprite(String sprite) {
        this.sprite = sprite;
    }

    public double getTopHeight() {
        return topHeight;
    }

    public int getLength() {
        return length;
    }

    public int getWidth() {
        return width;
    }

    public String getColour() {
        return colour;
    }

    public void setTopHeight(double topHeight) {
        this.topHeight = topHeight;
    }

    public String getBehaviourData() {
        return behaviourData;
    }

    public void setLength(int length) {
        this.length = length;
    }

    public void setWidth(int width) {
        this.width = width;
    }

    public List<ItemBehaviour> getBehaviourList() {
        return behaviourList;
    }
}

File: src/game/items/definition/item_definition.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <game/texts/external_texts_manager.h>

#include "item_definition.h"
#include "item_behaviour.h"
#include "util/stringbuilder.h"

item_definition *item_definition_create(int id, int cast_directory, char *sprite, char *colour, int length, int width, double top_height, char *behaviour) {
    item_definition *def = malloc(sizeof(item_definition));
    def->id = id;
    def->cast_directory = cast_directory;
    def->sprite = strdup(sprite);
    def->colour = strdup(colour);
    def->length = length;
    def->width = width;
    def->stack_height = top_height;
    def->top_height = top_height;
    def->behaviour_data = strdup(behaviour);
    def->behaviour = item_behaviour_parse(def);

    if (!def->behaviour->can_stack_on_top) {
        def->stack_height = 0;
    }

    if (def->stack_height == 0) {
        def->stack_height = 0.001;
    }

    return def;
}

item_definition *item_definition_create_blank() {
    item_definition *def = malloc(sizeof(item_definition));
    def->id = -1;
    def->cast_directory = -1;
    def->sprite = NULL;
    def->colour = NULL;
    def->length = 1;
    def->width = 1;
    def->stack_height = 0;
    def->top_height = 1;
    def->behaviour_data = NULL;
    def->behaviour = item_behaviour_parse(def);

    if (def->stack_height == 0) {
        def->stack_height = 0.001;
    }

    return def;
}

bool item_contains_custom_data(item_definition *definition) {
    return definition->behaviour->custom_data_numeric_on_off
            || definition->behaviour->custom_data_true_false
            || definition->behaviour->custom_data_numeric_state
            || definition->behaviour->custom_data_on_off
            || definition->behaviour->is_door;
}

/**
 * Get the item name by creating an external text key and reading external text entries.
 *
 * @param definition the item definition
 * @param special_sprite_id the special sprite id
 * @return the name
 */
char *item_definition_get_name(item_definition *definition, int special_sprite_id) {
    if (definition->behaviour->is_decoration) {
        return strdup(definition->sprite);
    } else {
        char *external_text_key = item_definition_get_text_key(definition, special_sprite_id);

        stringbuilder *sb = sb_create();
        sb_add_string(sb, external_text_key);
        sb_add_string(sb, "_name");

        char *name = texts_manager_get_value_by_id(sb->data);
        sb_cleanup(sb);

        if (name == NULL) {
            name = strdup("null");
        }

        free(external_text_key);
        return strdup(name);
    }
}

/**
 * Get the item description by creating an external text key and reading external text entries.
 *
 * @param definition the item definition
 * @param special_sprite_id the special sprite id
 * @return the description
 */
char *item_definition_get_desc(item_definition *definition, int special_sprite_id) {
    if (definition->behaviour->is_decoration) {
        return strdup(definition->sprite);
    } else {
        char *external_text_key = item_definition_get_text_key(definition, special_sprite_id);

        stringbuilder *sb = sb_create();
        sb_add_string(sb, external_text_key);
        sb_add_string(sb, "_desc");

        char *desc = texts_manager_get_value_by_id(sb->data);
        sb_cleanup(sb);

        if (desc == NULL) {
            desc = strdup("null");
        }

        free(external_text_key);
        return strdup(desc);
    }
}

/**
 * Create the catalogue icon through using the special sprite id.
 *
 * @param definition the item definition
 * @param special_sprite_id the special sprite id
 * @return the catalogue icon
 */
char *item_definition_get_icon(item_definition *definition, int special_sprite_id) {
    stringbuilder *sb = sb_create();
    sb_add_string(sb, definition->sprite);

    if (special_sprite_id > 0) {
        sb_add_string(sb, " ");
        sb_add_int(sb, special_sprite_id);
    }

    char *name = strdup(sb->data);
    sb_cleanup(sb);

    return name;
}


/**
 * Get external text key by definition.
 *
 * @param definition the item definition
 * @param special_sprite_id the special sprite id
 * @return the external text key
 */
char *item_definition_get_text_key(item_definition *definition, int special_sprite_id) {
    stringbuilder *sb = sb_create();

    if (special_sprite_id == 0) {
        if (definition->behaviour->is_wall_item) {
            sb_add_string(sb, "wallitem");
        } else {
            sb_add_string(sb, "furni");
        }

        sb_add_string(sb, "_");
    }

    sb_add_string(sb, definition->sprite);

    if (special_sprite_id > 0) {
        sb_add_string(sb, "_");
        sb_add_int(sb, special_sprite_id);
    }

    char *text_key = strdup(sb->data);
    sb_cleanup(sb);

    return text_key;
}

void item_definition_dispose(item_definition *def) {
    free(def->sprite);
    free(def->behaviour);
    free(def->behaviour_data);
    free(def->colour);
    free(def);
}